relay-server/src/services/processor/span.rs

```diff
@@ -16,16 +22,73 @@ pub use processing::*;
 
 pub fn filter(state: &mut ProcessEnvelopeState<SpanGroup>) {
     let disabled = state.should_filter(Feature::StandaloneSpanIngestion);
+    let otel_disabled = state.should_filter(Feature::OtelEndpoint);
+
     state.managed_envelope.retain_items(|item| {
-        if item.is_span() && disabled {
+        if disabled && item.is_span() {
             relay_log::debug!("dropping span because feature is disabled");
             ItemAction::DropSilently
+        } else if otel_disabled && item.ty() == &ItemType::OtelTracesData {
+            relay_log::debug!("dropping otel trace because feature is disabled");
+            ItemAction::DropSilently
         } else {
             ItemAction::Keep
         }
     });
 }
 
+pub fn convert_otel_traces_data(state: &mut ProcessEnvelopeState<SpanGroup>) {
+    let envelope = state.managed_envelope.envelope_mut();
+
+    for item in envelope.take_items_by(|item| item.ty() == &ItemType::OtelTracesData) {
+        convert_traces_data(item, &mut state.managed_envelope);
+    }
+}
+
+fn convert_traces_data(item: Item, managed_envelope: &mut TypedEnvelope<SpanGroup>) {
+    let traces_data = match parse_traces_data(item) {
+        Ok(traces_data) => traces_data,
+        Err(reason) => {
+            // NOTE: logging quantity=1 is semantically wrong, but we cannot know the real quantity
+            // without parsing.
+            track_invalid(managed_envelope, reason);
+            return;
+        }
+    };
+    for resource in traces_data.resource_spans {
+        for scope in resource.scope_spans {
+            for span in scope.spans {
+                // TODO: resources and scopes contain attributes, should denormalize into spans?
+                let Ok(payload) = serde_json::to_vec(&span) else {
+                    track_invalid(managed_envelope, DiscardReason::Internal);
+                    continue;
+                };
+                let mut item = Item::new(ItemType::OtelSpan);
+                item.set_payload(ContentType::Json, payload);
+                managed_envelope.envelope_mut().add_item(item);
```

Comment:
This is really unfortunate, but I can't think of anything better either ...

This brings up the question, is the `OtelSpan` item always Json, or also allowed to be protobuf?

Reply:
You mean the fact that we parse OTLP, serialize it as JSON, and then parse and convert it into our own span format again?

It's definitely suboptimal, but I prefer this over having different code paths for both.


> This brings up the question, is the OtelSpan item always Json, or also allowed to be protobuf?

Currently it has to be JSON:

https://github.com/getsentry/relay/blob/63a88f964098463277c70f0583446af8818d891c/relay-server/src/services/processor/span/processing.rs#L77-L78